lightning/src/routing/network_graph.rs

```diff
@@ -237,6 +282,349 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+					chain_hash.clone(),
+					Vec::with_capacity(short_channel_ids.len())
+				);
+				send_now = true;
+				query_scids_tasks.insert(their_node_id.clone(), task);
+				query_scids_tasks.get_mut(their_node_id).unwrap()
+			}
+		};
+
+		// Add the short_channel_ids to the task
+		query_scids_task.add(short_channel_ids);
+
+		// If this was a new task, then we should send it immediately.
+		// Otherwise, we already have a query in-flight and the new batch
+		// of scids will be sent when the in-flight query completes.
+		if send_now == true {
+			self.send_channels_query(query_scids_task);
+		}
+		Ok(())
+	}
+
+	fn handle_reply_channel_range(&self, their_node_id: &PublicKey, msg: &ReplyChannelRange) -> Result<(), LightningError> {
+		log_trace!(self.logger, "Handling reply_channel_range peer={}, first_blocknum={}, number_of_blocks={}, full_information={}, scids={}", log_pubkey!(their_node_id), msg.first_blocknum, msg.number_of_blocks, msg.full_information, msg.short_channel_ids.len(),);
+
+		// First we obtain the currently executing task for the peer
+		let mut query_range_tasks_lock = self.query_channel_range_tasks.lock().unwrap();
+		let query_range_tasks = &mut *query_range_tasks_lock;
+		let task = query_range_tasks.get_mut(their_node_id);
+
+		// If there is no currently executing task then we have received
+		// an invalid message and will return an error
+		if task.is_none() {
+			return Err(LightningError {
+				err: String::from("Received unknown reply_channel_range message"),
+				action: ErrorAction::IgnoreError,
+			});
+		}
+
+		let task = task.unwrap();
+
+		// Validate the chain_hash matches the chain_hash we used in the query.
+		// If it does not, then the message is malformed and we return an error
+		if msg.chain_hash != task.chain_hash {
+			query_range_tasks.remove(their_node_id);
+			return Err(LightningError {
+				err: String::from("Received reply_channel_range with invalid chain_hash"),
+				action: ErrorAction::IgnoreError,
+			});
+		}
+
+		// Validate that the remote node maintains up-to-date channel
+		// information for chain_hash. Some nodes use the full_information
+		// flag to indicate multi-part messages so we must check whether
+		// we received information as well.
+		if !msg.full_information && msg.short_channel_ids.len() == 0 {
+			query_range_tasks.remove(their_node_id);
+			return Err(LightningError {
+				err: String::from("Received reply_channel_range with no information available"),
+				action: ErrorAction::IgnoreError,
+			});
+		}
+
+		// On the first message...
+		if task.received_first_block.is_none() {
+			// Validate the first received message has a first_blocknum
+			// that is <= the start requested in our query which ensures
+			// the reply messages fully encompass our query If it does
+			// not, the message is malformed and we return an error.
+			if msg.first_blocknum > task.first_blocknum {
+				query_range_tasks.remove(their_node_id);
+				return Err(LightningError {
```

Comment:
Same, not covered by "Test receipt of a reply that has a first_blocknum that is above the first_blocknum requested in our query" ?

```
diff --git a/lightning/src/routing/network_graph.rs b/lightning/src/routing/network_graph.rs
index 8c98e0e8..f708a3e2 100644
--- a/lightning/src/routing/network_graph.rs
+++ b/lightning/src/routing/network_graph.rs
@@ -394,10 +394,11 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
                        // not, the message is malformed and we return an error.
                        if msg.first_blocknum > task.first_blocknum {
                                query_range_tasks.remove(their_node_id);
-                               return Err(LightningError {
-                                       err: String::from("Failing reply_channel_range with invalid first_blocknum"),
-                                       action: ErrorAction::IgnoreError,
-                               });
+                               return Ok(())
+                               //return Err(LightningError {
+                               //      err: String::from("Failing reply_channel_range with invalid first_blocknum"),
+                               //      action: ErrorAction::IgnoreError,
+                               //});
                        }
 
                        // Validate that the receive
```